File: src/input.py
#!/usr/bin/env python3

def ask_multiline_question(query: str) -> str:
    """
    Ask a question that can span multiple lines.
    The user enters lines until a blank line is entered.
    """
    print(query)
    lines = []
    while True:
        line = input()
        if not line.strip():  # Stop reading on an empty line
            break
        lines.append(line)
    return "\n".join(lines).strip()


def ask_question(query: str) -> str:
    """Ask a single-line question and return the answer."""
    return input(query)


def run():
    # 1. SET UP REPORT TOPIC
    # Get initial multi-line query from the user
    initial_query = ask_multiline_question(
        "What would you like to research? (Enter blank line to finish)"
    )

    # Get breadth parameter (with default of 4 if input is empty or invalid)
    breadth_input = ask_question("Enter research breadth (recommended 2-10, default 4): ")
    try:
        breadth = int(breadth_input)
    except ValueError:
        breadth = 4

    # Get depth parameter (with default of 2 if input is empty or invalid)
    depth_input = ask_question("Enter research depth (recommended 1-5, default 2): ")
    try:
        depth = int(depth_input)
    except ValueError:
        depth = 2

    print("Creating research plan...")
    print(initial_query)
    print(f"breadth={breadth}, depth={depth}")


if __name__ == "__main__":
    run()

File: src/input.ts
import * as fs from 'fs/promises';
import * as readline from 'readline';

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

//added a function so can ask long questions (Chatgpt https://chatgpt.com/c/67bf83e1-818c-800f-9151-a2d6bf8f0896?model=gpt-4o)
function askMultiLineQuestion(query: string): Promise<string> {
  return new Promise(resolve => {
    console.log(query);
    let response = "";
    rl.on("line", line => {
      if (!line.trim()) {  // Stop reading on empty line
        rl.removeAllListeners("line");
        resolve(response.trim());
      } else {
        response += (response ? "\n" : "") + line;
      }
    });
  });
}


// Helper function to get user input
function askQuestion(query: string): Promise<string> {
  return new Promise(resolve => {
    rl.question(query, answer => {
      resolve(answer);
    });
  });
}


// run the agent
// calls these major functions: askQuestion, generateFeedback, deepResearch, writeFinalReport, writeFile
async function run() {

  // 1. SET UP REPORT TOPIC
  // Get initial query
  // replace the original askQuestion function with Multiline-- it is believed new line characters cause bugs.
  // const initialQuery = await askQuestion('What would you like to research? ');
  const initialQuery = await askMultiLineQuestion("What would you like to research? (Enter blank line to finish)");

  // Get breath and depth parameters
  const breadth =
    parseInt(
      await askQuestion(
        'Enter research breadth (recommended 2-10, default 4): ',
      ),
      10,
    ) || 4;
  const depth =
    parseInt(
      await askQuestion('Enter research depth (recommended 1-5, default 2): '),
      10,
    ) || 2;

  log(`Creating research plan...`);
  log(`${initialQuery} `)
  log(`breadth=${breadth}, depth=${depth} `)
  rl.close();
}

run().catch(console.error);
